### PR TITLE
feat: support storing metadata in unixfs nodes

### PR DESCRIPTION
**Pull request by achingbrain**
Adds `mtime` and `mode` properties to `{path, content}` import entries.

### CODE REVIEW

**Review by alanshaw — COMMENTED** [review]
At what point is the file data moved to a separate node?

**Comment by achingbrain** [comment]
File data will end up in a leaf node if the `reduceSingleLeafToSelf` option is false, same as before.

**Comment by alanshaw** [comment]
Do you think we should have a default like, do not reduce to self if mtime or mode are not the defaults and file size is beyond some threshold?

File: README.md [hunk 1]
```diff
@@ -108,7 +105,9 @@ The `import` function returns an async iterator takes a source async iterator th
 ```js
 {
   path: 'a name',
-  content: (Buffer or iterator emitting Buffers)
+  content: (Buffer or iterator emitting Buffers),
+  mtime: (Number representing seconds since (positive) or before (negative) the Unix Epoch),
+  mode: (Number representing ugo-rwx, setuid, setguid and sticky bit)
```
Comment (alanshaw) [inline README.md:110]: Add a note that this is only applicable/persisted for files not directories?

File: src/dag-builder/index.js [hunk 2]
```diff
@@ -33,10 +33,18 @@ async function * dagBuilder (source, ipld, options) {
...
+        mode: entry.mode,
+        mtime: entry.mtime
+      })
     } else {
       // item is a directory
-      yield dirBuilder(entry, ipld, options)
+      yield dirBuilder(entry, ipld, {
+        ...options,
+        mode: entry.mode,
+        mtime: entry.mtime
```
Comment (alanshaw) [inline src/dag-builder/index.js]: Is it correct that these are not used by the `dirBuilder` but are passed to any files linked to from the directory?

File: src/dag-builder/index.js [hunk 3]
```diff
@@ -33,10 +33,18 @@ async function * dagBuilder (source, ipld, options) {
       const chunker = createChunker(options.chunker, validateChunks(source), options.chunkerOptions)
 
       // item is a file
-      yield fileBuilder(entry, chunker, ipld, options)
+      yield fileBuilder(entry, chunker, ipld, {
+        ...options,
+        mode: entry.mode,
+        mtime: entry.mtime
```
Comment (alanshaw) [inline src/dag-builder/index.js]: Why are these passed as options? - The entry is passed also.

File: test/importer.spec.js [hunk 4]
```diff
@@ -676,5 +676,43 @@ strategies.forEach((strategy) => {
...
+      for await (const file of importer([{
+        path: '1.2MiB.txt',
+        content: bigFile,
+        mode
+      }], ipld, options)) {
+        const node = await exporter(file.cid, ipld)
+
+        expect(node.unixfs.mode).to.equal(mode)
+      }
+    })
```
Comment (alanshaw) [inline test/importer.spec.js:843]: Test that a given file gets the default mode and mtime if unspecified?

Test you can add multiple files and they can each have a different mtime/mode and also have the defaults?

Test that you can't add meta to a directory?

Test that the meta isn't inherited by files in deeper directories?

**Review by alanshaw — CHANGES_REQUESTED** [review]

**Comment by alanshaw** [comment]
This is AWESOME btw!

**Review by hugomrdias — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (achingbrain) [reply README.md:110]: This is an oversight, metadata should be applied to directories too

File: src/dag-builder/index.js (hunk 2 above)
Comment (achingbrain) [reply src/dag-builder/index.js]: No, just the jetlag ~talking~ typing

File: src/dag-builder/index.js (hunk 3 above)
Comment (achingbrain) [reply src/dag-builder/index.js]: Fixed now

File: test/importer.spec.js (hunk 4 above)
Comment (achingbrain) [reply test/importer.spec.js:843]: Moar tests have been added

**Comment by achingbrain** [comment]
> Do you think we should have a default like, do not reduce to self if mtime or mode are not the defaults and file size is beyond some threshold?

Maybe, I'm not sure we should make decisions like that for the user, though I could be talked round.

One thing is that in order to work out the size of the node we'd need to serialize it first, then we serialize again inside IPLD when writing it to disk.  We already do this to get the size of `DAGNode`s in order to create `DAGLink`s so it'd make everything slower to do this in more places.

Something I'd like to try is removing the `.dag` invocations and using the repo directly for `dag-pb` operations, at that point we control serialization and hashing in one place and can optimise for the UnixFS use case which doesn't fit well with other IPLD types, mostly because of the `DAGLink` size property requirement above.

Anyway, I'd like to resolve that in a separate PR to this one, but one that gets resolved before this goes out of the door in a js-IPFS release.

**Comment by alanshaw** [comment]
> Maybe, I'm not sure we should make decisions like that for the user, though I could be talked round.

I think we should provide good defaults. The majority of users are not going to know or care about this so we should make a effort to ensure their datas get good deduplications out of the box.

We don't have to propose in this PR, but it would be good to get a follow up proposal PR so we can start considering it.

Also on the table, just changing the default chunker to rabin or the new buzhash chunker.

**Review by alanshaw — APPROVED** [review]

**Pull request merged**